base/namedtuple.jl

```diff
@@ -0,0 +1,91 @@
+# This file is a part of Julia. License is MIT: https://julialang.org/license
+
+@generated function namedtuple(::Type{NamedTuple{names,T} where T}, args...) where names
+    N = length(names)
+    if length(args) == N
+        Expr(:new, :(NamedTuple{names,$(Tuple{args...})}), Any[ :(args[$i]) for i in 1:N ]...)
+    else
+        :(throw(ArgumentError("wrong number of arguments to named tuple constructor")))
+    end
+end
+
+NamedTuple() = namedtuple(NamedTuple{()})
+
+length(t::NamedTuple) = nfields(t)
+start(t::NamedTuple) = 1
+done(t::NamedTuple, iter) = iter > nfields(t)
+next(t::NamedTuple, iter) = (getfield(t, iter), iter + 1)
+endof(t::NamedTuple) = length(t)
+getindex(t::NamedTuple, i::Int) = getfield(t, i)
+getindex(t::NamedTuple, i::Symbol) = getfield(t, i)
+
+function getindex(t::NamedTuple, I::AbstractVector)
+    names = unique( Symbol[ isa(i,Symbol) ? i : fieldname(typeof(t),i) for i in I ] )
```

Comment:
So this `getindex` method is kind of a bummer; I played around with a definition like:
```julia
@generated function Base.getindex(nt::NamedTuple{names}, inds::Int...) where {names}
    nms = [:(names[inds[$i]]) for i = 1:length(inds)]
    args = [:(getfield(nt, inds[$i])) for i = 1:length(inds)]
    return :(Base.namedtuple(NamedTuple{tuple($(nms...))}, $(args...)))
end
```
But the problem is you still don't know the values of `inds`, so it has a hard time knowing the exact return type for the NamedTuple. Is there any other way we could get a smarter multi-`getfield`-like type method?